pallets/compliance-manager/src/lib.rs

```diff
@@ -302,6 +302,8 @@ decl_module! {
 
             ensure!(Self::is_owner(&ticker, did), Error::<T>::Unauthorized);
 
+            ensure!(Self::get_latest_rule_id(ticker) >= asset_rule_id, Error::<T>::InvalidRuleId);
```

Comment:
This doesn't cover the case:
```rust
for _ in 0..2 {
    add_active_rule(...);
}
remove_active_rule(0); // OK; latest == 2
remove_active_rule(0); // BAD OK; latest == 2, but 0 was just removed.
```

You'd need to check for containment instead of length to capture the intention of "does id exist".
As an optimization you can alter `::mutate` below into `::try_mutate` and leverage the `.retain` itself for the containment check (size before/after, if changed, something was removed, hence success, otherwise failure.)

I'd also add test cases, including the one I've noted here.

Reply:
Just adding some more context here. The rule ids are not always consecutive, there can be gaps between the rule ids. We can have a situation where 5 rules are present and their rule ids are 1,5,10,11,12. To verify that a rule id exists, we need to iterate over the rules and check every rule id.

Unrelated - we can convert AssetRules into a double map where the second key is the id. This will allow us to do O(1) access. I'll add it as a different task in Jira.

Reply:
> Unrelated - we can convert AssetRules into a double map where the second key is the id. This will allow us to do O(1) access. I'll add it as a different task in Jira.

I believe there's an existing TODO in the code for this task (on `get_latest_rule_id`) but cannot hurt to user-story-ify it.